### DISCUSSION TITLE
Need to understand the scheduler

### DISCUSSION THREAD

**Discussion by shenke-adsk** [mend/renovate-ce-ee#409] [Q&A]
We run the Renovate CE container in our company infrastructure. One question we frequently get is how Renovate is scheduled to process the installations. To be more specific, the questions are:

Say we use the default schedule of every hour on the hour, if there is a large repository for which Renovate takes longer than 1 hour to process, what happens if the next scheduled run starts? Will the current Renovate job get terminated, or will it continue until finish? For later, will another Renovate job start on the same repository?

**Comment by JoeJennings** [comment]
I haven't moved to Renovate CE yet, but with Renovate OnPrem the job continues until it completes even if it takes 2+ hours.  Our source code management system is GitHub Enterprise and the authenticity tokens Renovate retrieves are only valid for 1 hour so Renovate completes processing only to have the PR creation fail because the token it has is no longer valid.

**Reply by shenke-adsk** [reply]
We use GHE as well. The 1-hour validity of the token is concerning. Is there any workaround? Like does Renovate have the ability to cache the change and on next run attempts to retry the PR?

**Comment by rarkins** [comment]
Renovate processes one repository at a time. New jobs don't start until the existing job finishes, regardless of scheduling.

The same token is used for the duration of the repo. GitHub app tokens are limited to one hour duration in length.

You should be enabling the repositoryCache option to reduce runtime per repository and speed up new runs.

**Reply by shenke-adsk** [reply]
> New jobs don't start until the existing job finishes, regardless of scheduling.

Can elaborate this further? So we use Renovate CE image located at `ghcr.io/mend/renovate-ce:6.7.0-full`. Is it like the following: 

1. The container starts, some scheduler process runs. 
2. The scheduler process runs based on the schedule defined by `MEND_RNV_CRON_JOB_SCHEDULER`. 
3. The scheduler reads from GHE what repositories have Renovate installed, and schedules each repository in a queue. 
4. Some worker takes one repository at a time. It invokes the Renovate CLI on it. 
5. If a repository takes a long time to run, the job will keep on running until finished or timed out. During this time, the scheduler may  run another round and new jobs are added to the queue? 

Also, where is the code for the scheduler? I don't see any source files in this repository. Or should I just shell into a running container and look at the files under `/usr/src/app/src`?

**Reply by rarkins** [reply]
Yes, you got the 5 steps right.

Renovate CE is not open source.

**Reply by shenke-adsk** [reply]
Got it. Any plan to also put it on the common Docker Hub (https://hub.docker.com/)? It used to be there when it was Renovate On-Prem. Or using GitHub Container Registry will become the common practice going forward?

**Reply by rarkins** [reply]
We plan to publish only to GHCR

**Comment by shenke-adsk** [comment]
I got what I wanted to know.
